### ISSUE TITLE
Rule Change: extend complexity rule to count optional chaining and default parameters/values

### ISSUE THREAD

**Issue by lo1tuma** [eslint/eslint#18060]
### What rule do you want to change?

complexity

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/*eslint complexity: ["error", 2]*/

function a(x) {
   return x?.a?.b?.c?.d?.e?.f;
}

function b(x) {
  const { a = '', b = 0 } = x;
} 

function c(a = '', b = 0) {
}
```

### What does the rule currently do for this code?

The rule doesn’t complain about increased complexity.

### What will the rule do after it's changed?

It will complain about all 3 functions that the complexity threshold got exceeded.

### Participation

- [X] I am willing to submit a pull request to implement this change.

**Comment by amareshsm** [comment]
I don't think optional chaining and parameters will increase the cyclomatic complexity of the code
```
function a(x) {
   return x?.a?.b?.c?.d?.e?.f;
}

```
This code doesn't have control flow statements such as loops or conditionals. It's a simple expression that uses the optional chaining operator (?.) to access nested properties in the object x. Therefore, the control flow graph is straightforward (ie 1) and doesn't introduce additional complexity.

```
/*eslint complexity: ["error", 2]*/

function a(x) {
   return x?.a?.b?.c?.d?.e?.f;
}

function b(x) {
  const { a = '', b = 0 } = x;
} 

function c(a = '', b = 0) {
}

```

According to [JSHint](https://jshint.com/), the complexity of the above functions is 1.

**Comment by lo1tuma** [comment]
> This code doesn't have control flow statements such as loops or conditionals. It's a simple expression that uses the optional chaining operator (?.)

How would you implement the code without optional chaining? I’m pretty sure that you would need a conditional. IMHO optional chaining is only syntax sugar for an if statement or an ternary. Every access to an optional property via optional chaining creates two possible code branches. Which is AFAIK the definition of cyclomatic complexity. In the example `x?.b` there are two possible branches in which the code can execute.

Also coverage tools like `c8` report untested optional chaining expressions as an uncovered branch.

> According to [JSHint](https://jshint.com/), the complexity of the above functions is 1.

I’m not sure if this is relevant. I’m not following the development of JSHint anymore, but I would be surprised if they implement rule-specific logic for modern JS language features.

**Comment by nzakas** [comment]
I agree with @lo1tuma -- both optional chaining and default parameters represent branches along a code path, which is the definition of complexity.

I'm :+1: for this change but would like some other opinions from @eslint/eslint-team.

**Comment by bmish** [comment]
Makes sense to count these. I'm used to test coverage tools counting them as separate branches.

**Comment by screendriver** [comment]
The null coalescing operator `??` should also be counted in my opinion because it also creates conditional branches which are increasing complexity:

- `value` is `null` or `undefined`
- `value` is not `null` or `undefined`

**Comment by JoshuaKGoldberg** [comment]
Agreement from me on account of what nzakas and bmish have already said.

**Comment by nzakas** [comment]
Looks like there is consensus, so marking as accepted. @lo1tuma if you want to submit a PR, we can review it.
